source/agora/consensus/data/DataPayload.d

```diff
@@ -0,0 +1,215 @@
+/*******************************************************************************
+
+    Contains a type for storing data in a transaction
+
+    Copyright:
+        Copyright (c) 2020 BOS Platform Foundation Korea
+        All rights reserved.
+
+    License:
+        MIT License. See LICENSE for details.
+
+*******************************************************************************/
+
+module agora.consensus.data.DataPayload;
+
+import agora.common.Hash;
+import agora.common.Serializer;
+
+import std.algorithm.iteration;
+import std.conv: parse;
+import std.range;
+
+/// The structure of for storing data in a transaction
+public struct DataPayload
+{
+    /// The byte array of transaction data
+    public ubyte[] data;
+
+    /***************************************************************************
+
+        Create a DataPayload from binary data
+
+        Params:
+            bin  = Binary data to store in this `DataPayload`
+
+    ***************************************************************************/
+
+    public this (scope const ubyte[] bin) pure nothrow @safe
+    {
+        this.data.length = bin.length;
+        this.data[] = bin[];
+    }
```

Comment:
How about changing the data to `const`? After all, the data should never be modified after its read. And this way you avoid having to duplicate the array:

```suggestion
    /// The byte array of transaction data
    public const(ubyte)[] data;

    /***************************************************************************

        Create a DataPayload from binary data

        Params:
            bin  = Binary data to store in this `DataPayload`

    ***************************************************************************/

    public this (const(ubyte)[] bin) pure nothrow @safe
    {
        this.data.bin;
    }
```

Reply:
That's a good amendment. I changed it.